File: python/tests/unit/test_bit_activity.py
import matplotlib.pyplot as plt

from adctoolbox.dout.bit_activity import bit_activity
from tests._utils import save_variable, save_fig
from tests.unit._runner import run_unit_test_batch

plt.rcParams['font.size'] = 16
plt.rcParams['axes.grid'] = True

def _process_bit_activity(raw_data, sub_folder, dataset_name):
    """
    Callback function to process a single file:
    1. Run bit activity analysis
    2. Save bit usage variable
    3. Save plot
    """
    # Create figure and run bit_activity
    fig = plt.figure(figsize=(10, 7.5))
    bit_usage = bit_activity(raw_data, annotate_extremes=True)
    plt.gca().tick_params(labelsize=16)    
    plt.title(f'Bit activity: {dataset_name}')
    
    save_fig(sub_folder, 'bitActivity.png', dpi=150)

    # Save bit_usage data
    save_variable(sub_folder, bit_usage, 'bit_usage')

def test_bit_activity(project_root):
    """
    Batch runner for bit activity analysis.
    """
    run_unit_test_batch(
        project_root=project_root,
        input_subpath="dataset/dout",
        test_module_name="test_bit_activity",
        file_pattern="dout_*.csv",
        output_subpath="test_output",
        process_callback=_process_bit_activity,
        flatten=False  # Digital output data is 2D (N samples x M bits)
    )

File: matlab/tests/unit/test_bitActivity.m
%% test_bitActivity.m
close all; clc; clear;

%% Configuration
verbose = 0;
inputDir = "dataset/dout";
outputDir = "test_output";
filesList = {};
filesList = autoSearchFiles(filesList, inputDir, 'dout_*.csv');
if ~isfolder(outputDir), mkdir(outputDir); end

%% Test Loop
for k = 1:length(filesList)
    currentFilename = filesList{k};
    dataFilePath = fullfile(inputDir, currentFilename);
    fprintf('[%s] [%d/%d] [%s]\n', mfilename, k, length(filesList), currentFilename);

    bits = readmatrix(dataFilePath);

    figure('Position', [100, 100, 800, 600], "Visible", verbose);
    bit_usage = bitActivity(bits, 'AnnotateExtremes', true);
    set(gca, "FontSize", 16);

    [~, datasetName, ~] = fileparts(currentFilename);
    subFolder = fullfile(outputDir, datasetName, mfilename);
    saveFig(subFolder, "bitActivity_matlab.png", verbose);
    saveVariable(subFolder, bit_usage, verbose);
end
